datadog_checks_base/datadog_checks/base/utils/db/utils.py

```diff
@@ -278,3 +278,71 @@ def _run_job_rate_limited(self):
 
     def run_job(self):
         raise NotImplementedError()
+
+
+def dbm_tracked_method(dbms, parent_check_attr=None, track_result_length=False):
+    """
+    Decorates a function to provide debug metrics and logging for troubleshooting.
+    Tracks execution time, errors, and result length.
+
+    If the check has a `debug_stats_kwargs` function, that function is called to get a set of kwargs to pass to
+    the statsd methods (i.e. histogram, count, gauge, etc). This is useful when specific tags need to be added to
+    these debug metrics in a standardized way.
+
+    :param dbms: the dbms to report as (i.e. sqlserver)
+    :param parent_check_attr: the attribute of the class which references the check. If none, self must refer to the
+    check. The check reference is used to report debug metrics and to generate logs.
+    :param track_result_length: if true, the length of the result is tracked
+    :return: a decorator
+    """
+
+    def decorator(function):
+        def wrapper(self, *args, **kwargs):
+            start_time = time.time()
+
+            try:
+                check = getattr(self, parent_check_attr) if parent_check_attr else self
+            except AttributeError:
+                print("[{} {}] invalid dbm_tracked_method. invalid check reference".format(dbms, function.__name__))
+                return function(self, *args, **kwargs)
+
+            stats_kwargs = {}
+            if hasattr(check, 'debug_stats_kwargs'):
+                stats_kwargs = dict(check.debug_stats_kwargs())
+
+            stats_kwargs['tags'] = stats_kwargs.get('tags', []) + ["operation:{}".format(function.__name__)]
+
+            try:
+                result = function(self, *args, **kwargs)
+
+                elapsed_ms = (time.time() - start_time) * 1000
+                check.histogram(
+                    "dd.{}.operation.time".format(dbms),
+                    elapsed_ms,
+                    **stats_kwargs,
+                )
+
+                check.log.debug("[%s %s] operation completed in %s ms", elapsed_ms)
+
+                if track_result_length and result is not None:
```

Comment:
This seems like a weird metric since not everything supports `len` and also doesn't give an indication of nested sizes. Maybe `sys.getsizeof` is better?

Reply:
Yeah but it's off by default so you'd enable it only on methods where you known you're returning a list-like thing and want to track the length of that. The example is that several of the methods we have that read out lists of results from the database would benefit from being able to track the response length in an easy way.

Reply:
`sys.getsizeof` is not the same as the length of the list so it would give different information.